File: src/go/upack/command.go
package upack

import (
	"archive/zip"
	"crypto/sha1"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"github.com/pkg/errors"
)

type Command interface {
	Name() string
	Description() string

	Help() string
	Usage() string

	PositionalArguments() []PositionalArgument
	ExtraArguments() []ExtraArgument

	Run() int
}

type PositionalArgument struct {
	Index       int
	Name        string
	Description string
	TrySetValue func(Command, *string) bool
	Optional    bool
}

type ExtraArgument struct {
	Name        string
	Alias       []string
	Description string
	TrySetValue func(Command, *string) bool
	Required    bool
	Flag        bool
}

func (a PositionalArgument) Help() string {
	return a.Name + " - " + a.Description
}

func (a ExtraArgument) Help() string {
	return a.Name + " - " + a.Description
}

func trySetBoolValue(name string, f func(Command) *bool) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil || *value == "" {
			*f(cmd) = true
			return true
		}

		if strings.EqualFold(*value, "true") {
			*f(cmd) = true
			return true
		}

		if strings.EqualFold(*value, "false") {
			*f(cmd) = false
			return true
		}
		fmt.Println("--"+name, "must be \"true\" or \"false\".")
		return false
	}
}

func trySetStringValue(name string, f func(Command) *string) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil {
			return false
		}

		*f(cmd) = *value
		return true
	}
}

func trySetStringFnValue(name string, f func(Command) func(string)) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil {
			return false
		}

		f(cmd)(*value)
		return true
	}
}

func trySetPathValue(name string, f func(Command) *string) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil {
			return false
		}

		p, err := filepath.Abs(*value)
		if err != nil {
			fmt.Println("--"+name, "must be a valid path.")
			return false
		}

		*f(cmd) = p
		return true
	}
}

func trySetBasicAuthValue(name string, f func(Command) **[2]string) func(Command, *string) bool {
	return func(cmd Command, value *string) bool {
		if value == nil || *value == "" {
			*f(cmd) = nil
			return true
		}

		parts := strings.SplitN(*value, ":", 2)
		if len(parts) != 2 {
			fmt.Println("--"+name, "must be in the format \"username:password\".")
			return false
		}

		*f(cmd) = &[2]string{parts[0], parts[1]}
		return true
	}
}

func (a PositionalArgument) Usage() string {
	s := "«" + a.Name + "»"

	if a.Optional {
		s = "[" + s + "]"
	}

	return s
}

func (a ExtraArgument) Usage() string {
	if !a.Required && a.Flag {
		return "[--" + a.Name + "]"
	}

	s := "--" + a.Name + "=«" + a.Name + "»"

	if !a.Required {
		s = "[" + s + "]"
	}

	return s
}

func defaultCommandUsage(cmd Command) string {
	s := []byte("upack ")

	s = append(s, cmd.Name()...)

	for _, arg := range cmd.PositionalArguments() {
		s = append(s, ' ')
		s = append(s, arg.Usage()...)
	}

	for _, arg := range cmd.ExtraArguments() {
		s = append(s, ' ')
		s = append(s, arg.Usage()...)
	}

	return string(s)
}

func defaultCommandHelp(cmd Command) string {
	s := []byte("Usage: ")

	s = append(s, cmd.Usage()...)
	s = append(s, '\n', '\n')
	s = append(s, cmd.Description()...)
	s = append(s, '\n')

	for _, arg := range cmd.PositionalArguments() {
		s = append(s, '\n')
		s = append(s, arg.Help()...)
	}

	for _, arg := range cmd.ExtraArguments() {
		s = append(s, '\n')
		s = append(s, arg.Help()...)
	}

	return string(s)
}

func ReadManifest(r io.Reader) (*UniversalPackageMetadata, error) {
	var meta UniversalPackageMetadata
	err := json.NewDecoder(r).Decode(&meta)
	if err != nil {
		return nil, err
	}
	return &meta, nil
}

func PrintManifest(info *UniversalPackageMetadata) {
	fmt.Println("Package:", info.groupAndName())
	fmt.Println("Version:", info.Version())
}

func UnpackZip(targetDirectory string, overwrite bool, zipFile *zip.Reader, preserveTimestamps bool) error {
	err := os.MkdirAll(targetDirectory, 0777)
	if err != nil {
		return err
	}

	var files int
	var directories int

	for _, entry := range zipFile.File {
		if !strings.HasPrefix(strings.ToLower(entry.Name), "package/") {
			continue
		}

		targetPath := filepath.Join(targetDirectory, entry.Name[len("package/"):])

		if entry.Mode().IsDir() {
			err = os.MkdirAll(targetPath, 0777)
			if err != nil {
				return err
			}
			fi, err := os.Stat(targetPath)
			if err != nil {
				return err
			}
			// Honor umask and make sure directory execute is set if directory read is set.
			mode := (entry.Mode() | (entry.Mode()&0444)>>2) & fi.Mode()
			err = os.Chmod(targetPath, mode)
			if err != nil {
				return err
			}

			directories++
		} else {
			err = os.MkdirAll(filepath.Dir(targetPath), 0777)
			if err != nil {
				return err
			}
			err = saveEntryToFile(entry, targetPath, overwrite, preserveTimestamps)
			if err != nil {
				return err
			}

			files++
		}
	}

	fmt.Println("Extracted", files, "files and", directories, "directories.")
	return nil
}

func saveEntryToFile(entry *zip.File, targetPath string, overwrite, preserveTimestamps bool) (err error) {
	r, err := entry.Open()
	if err != nil {
		return
	}
	defer func() {
		if e := r.Close(); err == nil {
			err = e
		}
	}()

	flags := os.O_WRONLY | os.O_TRUNC | os.O_CREATE
	if !overwrite {
		flags |= os.O_EXCL
	}

	f, err := os.OpenFile(targetPath, flags, entry.Mode())
	if err != nil {
		return
	}
	defer func() {
		if e := f.Close(); err == nil {
			err = e
		}
	}()

	_, err = io.Copy(f, r)
	if err != nil {
		return
	}

	if preserveTimestamps && entry.Modified.Year() > 1980 {
		err = os.Chtimes(targetPath, entry.Modified, entry.Modified)
		if err != nil {
			return
		}
	}

	return
}

func CreateEntryFromFile(zipFile *zip.Writer, fileName, entryPath string) (err error) {
	f, err := os.Open(fileName)
	if err != nil {
		return
	}
	defer func() {
		if e := f.Close(); err == nil {
			err = e
		}
	}()

	fi, err := f.Stat()
	if err != nil {
		return
	}

	h, err := zip.FileInfoHeader(fi)
	if err != nil {
		return
	}

	h.Name = entryPath
	h.Method = zip.Deflate

	w, err := zipFile.CreateHeader(h)
	if err != nil {
		return
	}

	_, err = io.Copy(w, f)
	return
}

func CreateEntryFromStream(zipFile *zip.Writer, file io.Reader, entryPath string) (err error) {
	w, err := zipFile.Create(entryPath)
	if err != nil {
		return
	}

	_, err = io.Copy(w, file)
	return
}

func AddDirectory(zipFile *zip.Writer, sourceDirectory, entryRootPath string) (err error) {
	fi, err := os.Stat(sourceDirectory)
	if err != nil {
		return
	}

	h, err := zip.FileInfoHeader(fi)
	if err != nil {
		return
	}

	h.Name = entryRootPath

	_, err = zipFile.CreateHeader(h)
	if err != nil {
		return
	}

	infos, err := ioutil.ReadDir(sourceDirectory)
	if err != nil {
		return
	}
	for _, fi := range infos {
		if fi.IsDir() {
			err = AddDirectory(zipFile, filepath.Join(sourceDirectory, fi.Name()), entryRootPath+fi.Name()+"/")
		} else {
			err = CreateEntryFromFile(zipFile, filepath.Join(sourceDirectory, fi.Name()), entryRootPath+fi.Name())
		}

		if err != nil {
			return
		}
	}

	return
}

func GetVersion(source, group, name, version string, credentials *[2]string, prerelease bool) (string, error) {
	if version != "" && !strings.EqualFold(version, "latest") && !prerelease {
		return version, nil
	}

	req, err := http.NewRequest("GET", strings.TrimRight(source, "/")+"/packages?"+(url.Values{"group": {group}, "name": {name}}).Encode(), nil)
	if err != nil {
		return "", err
	}

	if credentials != nil {
		req.SetBasicAuth(credentials[0], credentials[1])
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("ProGet returned HTTP error: %s", resp.Status)
	}

	var data RemotePackageMetadata
	err = json.NewDecoder(resp.Body).Decode(&data)
	if err != nil {
		return "", err
	}

	if len(data.Versions) == 0 {
		groupAndName := name
		if group != "" {
			groupAndName = group + "/" + name
		}
		return "", fmt.Errorf("No versions of package %s found.", groupAndName)
	}

	var latestVersion *UniversalPackageVersion
	for _, v := range data.Versions {
		version, err := ParseUniversalPackageVersion(v)
		if err != nil {
			return "", err
		}
		if !prerelease && version.Prerelease != "" {
			continue
		}
		if latestVersion == nil || latestVersion.Compare(version) < 0 {
			latestVersion = version
		}
	}
	return latestVersion.String(), nil
}

func GetSHA1(filePath string) (h string, err error) {
	f, err := os.Open(filePath)
	if err != nil {
		return
	}
	defer func() {
		if e := f.Close(); err == nil {
			err = e
		}
	}()

	hash := sha1.New()
	_, err = io.Copy(hash, f)
	if err != nil {
		return
	}

	h = hex.EncodeToString(hash.Sum(nil))
	return
}

func GetPackageMetadata(packagePath string) (metadata *UniversalPackageMetadata, err error) {
	pkg, err := zip.OpenReader(packagePath)
	if err != nil {
		return nil, errors.Wrapf(err, "The source package '%s' does not exist or could not be opened.", packagePath)
	}
	defer func() {
		if e := pkg.Close(); err == nil {
			err = errors.Wrapf(e, "The source package '%s' does not exist or could not be opened.", packagePath)
		}
	}()

	for _, entry := range pkg.File {
		if entry.Name == "upack.json" {
			var r io.ReadCloser
			r, err = entry.Open()
			if err != nil {
				return nil, errors.Wrapf(err, "The source package '%s' does not exist or could not be opened.", packagePath)
			}
			defer func() {
				if e := r.Close(); err == nil {
					err = errors.Wrapf(e, "The source package '%s' does not exist or could not be opened.", packagePath)
				}
			}()

			metadata, err = ReadManifest(r)
			if err != nil {
				err = errors.Wrapf(err, "The source package '%s' does not exist or could not be opened.", packagePath)
			}
			return
		}
	}
	return nil, errors.Errorf("The source package '%s' does not exist or could not be opened.", packagePath)
}

func findChars(s string, f func(rune) bool) []string {
	var chars []string
	seen := make(map[rune]bool)

	for _, r := range s {
		if f(r) && !seen[r] {
			chars = append(chars, string(r))
			seen[r] = true
		}
	}

	return chars
}

func ValidateManifest(info *UniversalPackageMetadata) error {
	if info.Group() != "" {
		if len(info.Group()) > 250 {
			return errors.New("group must be between 0 and 250 characters long.")
		}

		invalid := findChars(info.Group(), func(c rune) bool {
			return (c < '0' || c > '9') && (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && c != '-' && c != '.' && c != '/' && c != '_'
		})
		if len(invalid) == 1 {
			return errors.New("group contains invalid character: '" + invalid[0] + "'")
		} else if len(invalid) > 1 {
			return errors.New("group contains invalid characters: '" + strings.Join(invalid, "', '") + "'")
		}

		if strings.HasPrefix(info.Group(), "/") || strings.HasSuffix(info.Group(), "/") {
			return errors.New("group must not start or end with a slash.")
		}
	}

	{
		if info.Name() == "" {
			return errors.New("missing name.")
		}
		if len(info.Name()) > 50 {
			return errors.New("name must be between 1 and 50 characters long.")
		}

		invalid := findChars(info.Name(), func(c rune) bool {
			return (c < '0' || c > '9') && (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && c != '-' && c != '.' && c != '_'
		})
		if len(invalid) == 1 {
			return errors.New("name contains invalid character: '" + invalid[0] + "'")
		} else if len(invalid) > 1 {
			return errors.New("name contains invalid characters: '" + strings.Join(invalid, "', '") + "'")
		}
	}

	_, err := ParseUniversalPackageVersion(info.Version())
	if err != nil {
		return errors.New("missing or invalid version.")
	}

	if len(info.Title()) > 50 {
		return errors.New("title must be between 0 and 50 characters long.")
	}

	return nil
}

File: src/dotnet/upack/Command.cs
using Inedo.UPack;
using Inedo.UPack.Net;
using Inedo.UPack.Packaging;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Net;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Inedo.ProGet.UPack
{
    public abstract class Command
    {
        [AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
        public sealed class PositionalArgumentAttribute : Attribute
        {
            public int Index { get; }
            public bool Optional { get; set; } = false;

            public PositionalArgumentAttribute(int index)
            {
                this.Index = index;
            }
        }

        [AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
        public sealed class ExtraArgumentAttribute : Attribute
        {
            public bool Optional { get; set; } = true;
        }

        [AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
        public sealed class ExpandPathAttribute : Attribute
        {
        }

        [AttributeUsage(AttributeTargets.Property, AllowMultiple = true, Inherited = true)]
        public sealed class AlternateNameAttribute : Attribute
        {
            public string Name { get; }

            public AlternateNameAttribute(string name)
            {
                this.Name = name;
            }
        }

        public abstract class Argument
        {
            protected readonly PropertyInfo p;

            internal Argument(PropertyInfo p)
            {
                this.p = p;
            }

            public abstract bool Optional { get; }
            public string DisplayName => p.GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? p.Name;
            public string Description => p.GetCustomAttribute<DescriptionAttribute>()?.Description ?? string.Empty;
            public object DefaultValue => p.GetCustomAttribute<DefaultValueAttribute>()?.Value;
            public bool ExpandPath => p.GetCustomAttribute<ExpandPathAttribute>() != null;

            public abstract string GetUsage();

            public virtual string GetHelp()
            {
                return $"{this.DisplayName} - {this.Description}";
            }

            public bool TrySetValue(Command cmd, string value)
            {
                if (p.PropertyType == typeof(bool))
                {
                    if (string.IsNullOrEmpty(value))
                    {
                        p.SetValue(cmd, true);
                        return true;
                    }
                    bool result;
                    if (bool.TryParse(value, out result))
                    {
                        p.SetValue(cmd, result);
                        return true;
                    }
                    Console.WriteLine($@"--{this.DisplayName} must be ""true"" or ""false"".");
                    return false;
                }

                if (p.PropertyType == typeof(string))
                {
                    if (this.ExpandPath)
                    {
                        string path = Path.GetFullPath(Path.Combine(Environment.CurrentDirectory, value ?? string.Empty));
                        p.SetValue(cmd, path);
                    }
                    else
                    {
                        p.SetValue(cmd, value);
                    }
                    return true;
                }

                if (p.PropertyType == typeof(NetworkCredential))
                {
                    if (string.IsNullOrWhiteSpace(value))
                    {
                        p.SetValue(cmd, null);
                        return true;
                    }

                    var parts = value.Split(new[] { ':' }, 2);
                    if (parts.Length != 2)
                    {
                        Console.WriteLine($@"--{this.DisplayName} must be in the format ""username:password"".");
                        return false;
                    }

                    p.SetValue(cmd, new NetworkCredential(parts[0], parts[1]));
                    return true;
                }

                throw new ArgumentException(p.PropertyType.FullName);
            }
        }

        public sealed class PositionalArgument : Argument
        {
            internal PositionalArgument(PropertyInfo p) : base(p)
            {
            }

            public int Index => p.GetCustomAttribute<PositionalArgumentAttribute>().Index;
            public override bool Optional => p.GetCustomAttribute<PositionalArgumentAttribute>().Optional;

            public override string GetUsage()
            {
                var s = $"«{this.DisplayName}»";

                if (this.Optional)
                {
                    s = $"[{s}]";
                }

                return s;
            }
        }

        public sealed class ExtraArgument : Argument
        {
            internal ExtraArgument(PropertyInfo p) : base(p)
            {
            }

            public IEnumerable<string> AlternateNames => p.GetCustomAttributes<AlternateNameAttribute>().Select(a => a.Name);
            public override bool Optional => p.GetCustomAttribute<ExtraArgumentAttribute>().Optional;

            public override string GetUsage()
            {
                var s = $"--{this.DisplayName}=«{this.DisplayName}»";

                if (this.Optional)
                {
                    s = $"[{s}]";
                }

                if (p.PropertyType == typeof(bool) && this.DefaultValue.Equals(false) && this.Optional)
                {
                    s = $"[--{this.DisplayName}]";
                }

                return s;
            }
        }

        public string DisplayName => this.GetType().GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? this.GetType().Name;
        public string Description => this.GetType().GetCustomAttribute<DescriptionAttribute>()?.Description ?? string.Empty;
        public IEnumerable<PositionalArgument> PositionalArguments => this.GetType().GetRuntimeProperties()
            .Where(p => p.GetCustomAttribute<PositionalArgumentAttribute>() != null)
            .Select(p => new PositionalArgument(p))
            .OrderBy(a => a.Index);

        public abstract Task<int> RunAsync(CancellationToken cancellationToken);

        public IEnumerable<ExtraArgument> ExtraArguments => this.GetType().GetRuntimeProperties()
            .Where(p => p.GetCustomAttribute<ExtraArgumentAttribute>() != null)
            .Select(p => new ExtraArgument(p));

        public string GetUsage()
        {
            var s = new StringBuilder("upack ");

            s.Append(this.DisplayName);

            foreach (var arg in this.PositionalArguments)
            {
                s.Append(' ').Append(arg.GetUsage());
            }

            foreach (var arg in this.ExtraArguments)
            {
                s.Append(' ').Append(arg.GetUsage());
            }

            return s.ToString();
        }

        public string GetHelp()
        {
            var s = new StringBuilder("Usage: ");

            s.AppendLine(this.GetUsage()).AppendLine().AppendLine(this.Description);

            foreach (var arg in this.PositionalArguments)
            {
                s.AppendLine().Append(arg.GetHelp());
            }

            foreach (var arg in this.ExtraArguments)
            {
                s.AppendLine().Append(arg.GetHelp());
            }

            return s.ToString();
        }

        internal static async Task<UniversalPackageMetadata> ReadManifestAsync(Stream metadataStream)
        {
            using (var reader = new StreamReader(metadataStream))
            {
                var text = await reader.ReadToEndAsync();
                return JsonConvert.DeserializeObject<UniversalPackageMetadata>(text);
            }
        }

        internal static string ValidateManifest(UniversalPackageMetadata info)
        {
            if (info.Group != null)
            {
                if (info.Group.Length > 250)
                {
                    return "group must be between 0 and 250 characters long.";
                }

                var invalid = info.Group.Where(c => (c < '0' || c > '9') && (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && c != '-' && c != '.' && c != '/' && c != '_').Distinct().ToArray();
                if (invalid.Length == 1)
                {
                    return "group contains invalid character: '" + invalid[0] + "'";
                }
                else if (invalid.Length > 1)
                {
                    return "group contains invalid characters: '" + string.Join("', '", invalid) + "'";
                }

                if (info.Group.StartsWith("/") || info.Group.EndsWith("/"))
                {
                    return "group must not start or end with a slash.";
                }
            }

            {
                if (string.IsNullOrEmpty(info.Name))
                {
                    return "missing name.";
                }
                if (info.Name.Length > 50)
                {
                    return "name must be between 1 and 50 characters long.";
                }

                var invalid = info.Name.Where(c => (c < '0' || c > '9') && (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && c != '-' && c != '.' && c != '_').Distinct().ToArray();
                if (invalid.Length == 1)
                {
                    return "name contains invalid character: '" + invalid[0] + "'";
                }
                else if (invalid.Length > 1)
                {
                    return "name contains invalid characters: '" + string.Join("', '", invalid) + "'";
                }
            }

            if (UniversalPackageVersion.TryParse(info["version"] as string ?? string.Empty) == null)
            {
                return "missing or invalid version.";
            }

            if (info.Title != null && info.Title.Length > 50)
            {
                return "title must be between 0 and 50 characters long.";
            }

            return null;
        }

        internal static void PrintManifest(UniversalPackageMetadata info)
        {
            if (!string.IsNullOrEmpty(info.Group))
            {
                Console.WriteLine($"Package: {info.Group}/{info.Name}");
            }
            else
            {
                Console.WriteLine($"Package: {info.Name}");
            }

            Console.WriteLine($"Version: {info.Version}");
        }

        internal static async Task UnpackZipAsync(string targetDirectory, bool overwrite, UniversalPackage package, bool preserveTimestamps, CancellationToken cancellationToken)
        {
            Directory.CreateDirectory(targetDirectory);

            var entries = package.Entries.Where(e => e.IsContent);

            int files = 0;
            int directories = 0;

            foreach (var entry in entries)
            {
                var targetPath = Path.Combine(targetDirectory, entry.ContentPath);

                if (entry.IsDirectory)
                {
                    Directory.CreateDirectory(targetPath);
                    directories++;
                }
                else
                {
                    Directory.CreateDirectory(Path.GetDirectoryName(targetPath));
                    using (var entryStream = entry.Open())
                    using (var targetStream = new FileStream(targetPath, overwrite ? FileMode.Create : FileMode.CreateNew, FileAccess.Write, FileShare.None, 4096, FileOptions.Asynchronous))
                    {
                        await entryStream.CopyToAsync(targetStream, 65536, cancellationToken);
                    }

                    // Assume files with timestamps set to 0 (DOS time) or close to 0 are not timestamped.
                    if (preserveTimestamps && entry.Timestamp.Year > 1980)
                    {
                        File.SetLastWriteTimeUtc(targetPath, entry.Timestamp.DateTime);
                    }

                    files++;
                }
            }

            Console.WriteLine($"Extracted {files} files and {directories} directories.");
        }

        internal static async Task<UniversalPackageVersion> GetVersionAsync(UniversalFeedClient client, UniversalPackageId id, string version, bool prerelease, CancellationToken cancellationToken)
        {
            if (!string.IsNullOrEmpty(version) && !string.Equals(version, "latest", StringComparison.OrdinalIgnoreCase) && !prerelease)
            {
                var parsed = UniversalPackageVersion.TryParse(version);
                if (parsed != null)
                    return parsed;

                throw new UpackException($"Invalid UPack version number: {version}");
            }

            IReadOnlyList<RemoteUniversalPackageVersion> versions;
            try
            {
                versions = await client.ListPackageVersionsAsync(id, false, null, cancellationToken);
            }
            catch (WebException ex)
            {
                throw ConvertWebException(ex);
            }

            if (!versions.Any())
                throw new UpackException($"No versions of package {id} found.");

            return versions.Max(v => v.Version);
        }

        internal const string PackageNotFoundMessage = "The specified universal package was not found at the given URL";
        internal const string FeedNotFoundMessage = "No UPack feed was found at the given URL";
        internal const string IncorrectCredentialsMessage = "The server rejected the username or password given";

        internal static UpackException ConvertWebException(WebException ex, string notFoundMessage = FeedNotFoundMessage)
        {
            var message = ex.Message;
            var statusCode = (ex.Response as HttpWebResponse)?.StatusCode;
            if (ex.Status == WebExceptionStatus.ProtocolError && statusCode.HasValue)
            {
                if (statusCode == HttpStatusCode.NotFound)
                {
                    message = notFoundMessage;
                }
                else if (statusCode == HttpStatusCode.Unauthorized)
                {
                    message = IncorrectCredentialsMessage;
                }

                if (ex.Response.ContentType == "text/plain")
                {
                    try
                    {
                        using (var reader = new StreamReader(ex.Response.GetResponseStream()))
                        {
                            var body = reader.ReadToEnd();
                            if (!string.IsNullOrWhiteSpace(body))
                            {
                                message = message + ": " + body;
                            }
                        }
                    }
                    catch
                    {
                        // ignore
                    }
                }
            }
            return new UpackException(message, ex);
        }

        internal static UniversalFeedClient CreateClient(string source, NetworkCredential credentials)
        {
            try
            {
                var uri = new Uri(source);

                var endpoint = credentials == null ?
                    new UniversalFeedEndpoint(uri, true) :
                    new UniversalFeedEndpoint(uri, credentials.UserName, credentials.SecurePassword);

                return new UniversalFeedClient(endpoint);
            }
            catch (UriFormatException ex)
            {
                throw new UpackException("Invalid UPack feed URL: " + ex.Message, ex);
            }
            catch (ArgumentException ex)
            {
                throw new UpackException("Invalid UPack feed URL: " + ex.Message, ex);
            }
        }

        internal static HexString GetSHA1(string filePath)
        {
            using (var file = File.OpenRead(filePath))
            using (var hash = HashAlgorithm.Create("SHA1"))
            {
                var bytes = hash.ComputeHash(file);
                return new HexString(bytes);
            }
        }

        internal static UniversalPackageMetadata GetPackageMetadata(string zipFileName)
        {
            try
            {
                using (var package = new UniversalPackage(zipFileName))
                {
                    return package.GetFullMetadata().Clone();
                }
            }
            catch (Exception ex)
            {
                throw new UpackException($"The source package file '{zipFileName}' does not exist or could not be opened.", ex);
            }
        }
    }
}
